src/routing_table/network_tests.rs

```diff
@@ -242,48 +242,45 @@ fn node_to_group_message() {
     }
 }
 
-fn verify_invariant(network: &mut Network) {
-    let keys = network.keys();
-    for _ in 0..20 {
-        let address = network.rng.gen();
-        let close_peer = network.close_node(address);
-        let group = unwrap!(network.nodes[&close_peer].close_names(&address));
-        for &node in &keys {
-            match network.nodes[&node].close_names(&address) {
-                None => assert!(!group.contains(&address)),
-                Some(nodes) => {
-                    if network.nodes[&node].is_recipient(&Destination::Group(address)) {
-                        assert_eq!(group, nodes);
-                    } else {
-                        for candidate in &group {
-                            assert!(nodes.contains(&candidate));
-                        }
-                    }
-                }
+fn verify_invariant(network: &Network) {
+    let mut groups: HashMap<Prefix<u64>, HashSet<u64>> = HashMap::new();
+    // first, collect all groups in the network
+    for node in network.nodes.values() {
+        for prefix in node.groups.keys() {
+            let mut group_content = node.groups[prefix].clone();
+            if *prefix == node.our_group_prefix {
+                group_content.insert(*node.our_name());
             }
+            if let Some(group) = groups.get_mut(prefix) {
+                group.extend(group_content);
+                continue;
+            }
+            let _ = groups.insert(*prefix, group_content);
         }
+        // use this opportunity to check if each node satisfies the invariant
+        node.verify_invariant();
     }
+    // check that prefixes are disjoint
+    verify_disjoint_prefixes(groups.keys().into_iter().cloned().collect());
+
+    // check that each group contains names agreeing with its prefix
+    verify_groups_match_names(&groups);
+
+    // check that groups cover the whole namespace
+    assert_eq!(network.nodes.len(), groups.values().map(|x| x.len()).sum());
```

Comment:
Ah, sorry, that's not quite what I meant: We need to assert that every address (not only of a group, but every possible value of `u64`) belongs to one of the groups.
Every group - with prefix `p` - corresponds to an interval of `u64` values: `[p00...00, p11...11]` - all values starting with that prefix. The union of these intervals needs to be the whole namespace `[00...00, 11...11]`.

I'm not sure what's the best way to assert that yet, though, so I'm happy to just add a `TODO` here and add this later.
One approach could be to sort the prefixes and assert that the intervals corresponding to subsequent groups always touch: The first interval must start with `0`, the last one must end with `11...11` (i.e. `2**64 - 1`), and for any subsequent intervals `[a, b]` and `[c, d]`, `c = b + 1`.